### PR TITLE
WIP: Sign Email

### PR DESCRIPTION
**Pull request by zeripath**
This provides a mechanism for signing email from the gitea server with a GPG key.

There is a slight inefficiency with emails sent to participants of an issue - whereby the signature is generated for each message in turn - because they get sent as individual messages.

Related #4398

### CODE REVIEW

**Comment by lafriks** [comment]
Wouldn't it be better to use https://godoc.org/golang.org/x/crypto/openpgp for this?

**Comment by zeripath** [comment]
We have no way of storing a private key to sign things inside Gitea at present. This just uses the host gpg similar to that used when git signs commits.

If we start encrypting mail then definitely I'll use the golang libraries.

File: services/mailer/mailer.go [hunk 1]
```diff
@@ -61,10 +101,13 @@ func (m *Message) ToMessage() *gomail.Message {
...
 			log.Warn("Mail contains HTML but configured to send as plain text.")
 		}
-		msg.SetBody("text/plain", plainBody)
+		msg.SetBody("text/plain", clearSignedMessage(plainBody))
 	} else {
-		msg.SetBody("text/plain", plainBody)
+		msg.SetBody("text/plain", clearSignedMessage(plainBody))
 		msg.AddAlternative("text/html", m.Body)
+		if setting.MailService.SignEmails {
+			msg.AddAlternative("application/pgp-signature", detachedSignature(m.Body))
```
Comment (guillep2k) [inline services/mailer/mailer.go:109]: The HTML version is not included in the signature. What does the user agent do in that case? Skip the HTML? 🤔

Comment (zeripath) [reply services/mailer/mailer.go:109]: The html block is signed with the detached signature

Comment (zeripath) [reply services/mailer/mailer.go:109]: I don't think the there's a way of signing both with one signature hence the decision to sign the plaintext in clear sign and the html as a detached signature. That being said if you know what the best practice is I'm only guessing here.

**Comment by zeripath** [comment]
I think I may need to redo this as a multipart/signed containing a multipart but I'm not sure.

**Comment by zeripath** [comment]
OK so it appears that gomail is dead.

We don't need a lot of its features anyway as we don't send attachments or embeded images, and we don't use its embedded dial either.

It can't provide mime multipart/signed or encrypted so I'm gonna rip it out. It's another dependency we don't need.

**Comment by zeripath** [comment]
I'm gonna close this as it needs a lot more thought - and testing data - to get right.

**Comment by alexanderadam** [comment]
Wow, it would have been a nice feature! So thank you for your efforts anyway!
 I'm crossing fingers :crossed_fingers: that someone implements this with a maintained dependency.

**Comment by zeripath** [comment]
It's not that gomail is the issue. I just couldn't find examples of what a mutlitpart signed messages looked like and how to do it. If you are using email encryption and signed emails could you send me a signed multipart email - that is one with plain text and html mail so I can copy it and build this feature.

**Pull request closed without merging**